build: Enable standalone build

DanielJerolm opened the pull request:
Until now GPCC was always build as a sub-project of a top-level CMake project. GPCC's unittest cases were build into an unittest executable defined by the top level project, maybe together with unittests from other components.

This PR shall add the capability to build GPCC standalone inclusive building a unittest executable. This shall pave the way to add github actions/workflows to build and run GPCC's unittest cases and sanitizers automatically in a subsequent PR.

If build standalone, only building for the host (native linux) shall be supported. This allows to get rid of the external dependency `toolchain_configs`. Building for other OS standalone will not be supported, because these scenarios usually require special toolchains that are not broadly available.

The following changes shall be made:
- No separate CMakeLists.txt-files any more for the target environments "productive" and "unittest"
- GPCC's CMakeList files shall detect if GPCC is included as a sub-project:  
  In this case there shall be no (major) breaking change, at best no change for the top-project at all.
- GPCC's CMakeList files shall detect if GPCC is build stand-alone. In thise case:
  - A unittest executable shall be build if the target environment is `unittest`.
  - All dependencies shall be fetched silently by CMake.
- In any case, dependency `toolchain configs` shall be eliminated.
- Scripts for triggering all build steps shall be provided.  
  The scripts provide examples and simplify handling.
- A .vscode folder with settings and tasks shall be added to enable editing source code in both target environments and to trigger the most relevant build-steps from VSCODE

These changes have been implemented until creation of this pull-request.

The following features are planned to be added after some review and discussion:
- Auto detection of configuration, if GPCC_OS and GPCC_Compiler are both not specified
- Update of CMake style for naming variables
- Generation of doxygen scripts by CMake if feasible

Until completion, this will be a draft PR.

DanielJerolm committed:
Wrap-up TODOs

DanielJerolm committed:
Add .vscode folder

DanielJerolm committed:
Revise doxygen scripts

DanielJerolm committed:
Revise CMake and build-scripts

DanielJerolm committed:
Enable automativ word-wrap in md-files

DanielJerolm committed:
Render names of build scripts more precise

DanielJerolm committed:
Update README.md

DanielJerolm committed:
Fix name of EPOS kernel lib

DanielJerolm committed:
Add CMakeLists.md

Wishmaster-de commented on CMakeLists.txt:
```diff
@@ -107,41 +43,138 @@ string(REPLACE ";"  ", " osValues_str "${GPCC_OSValues}")
+add_library(${PROJECT_NAME})
+
+if(${GPCC_TargetEnvironment} STREQUAL "productive")
+  add_subdirectory(src)
```
Maybe it's better to have only a ' // Intentional left free' here and put the line upper the if line.
Justification: In all cases add_subdirectory(src) will be added and there will be no duplication.

DanielJerolm replied:
Will be applied

Wishmaster-de reviewed (commented):
I tried it in standalone mode. Googletest was fetched. Building the unittests fails with

```terminal
PR-gpcc/gpcc/test_src/osal/tfc/PotentialUnreproducibleBehaviourTrap.cpp:15:10: fatal error: gtest/gtest.h: No such file or directory
   15 | #include <gtest/gtest.h>
      |          ^~~~~~~~~~~~~~~
compilation terminated.
PR-gpcc/gpcc/test_src/osal/tfc/BlockWithExpiredTimeoutTrap.cpp:15:10: fatal error: gtest/gtest.h: No such file or directory
   15 | #include <gtest/gtest.h>
      |          ^~~~~~~~~~~~~~~
compilation terminated.
PR-gpcc/gpcc/test_src/osal/tfc/UnreproducibleBehaviourTrap.cpp:15:10: fatal error: gtest/gtest.h: No such file or directory
   15 | #include <gtest/gtest.h>
```

If i changed -DGPCC_OS:STRING=linux_x64_tfc to -DGPCC_OS:STRING=linux_x64 in file cmake_config_nativelinux-unittest-debug.sh building succeeds. :astonished:

DanielJerolm committed:
Fix review finding: Code dublication

DanielJerolm committed:
Add missing dependency from gpcc to googletest in unittest env

DanielJerolm commented:
@Wishmaster-de , I cannot reproduce the issue on my system.

But I think this is the reason: There is a dependency from library `gpcc` to `googletest` when building for the unittest environment. That dependency was not specified by a `target_link_libraries()` command.

I commited a fix.

Wishmaster-de commented:
> @Wishmaster-de , I cannot reproduce the issue on my system.
> 
> But I think this is the reason: There is a dependency from library `gpcc` to `googletest` when building for the unittest environment. That dependency was not specified by a `target_link_libraries()` command.
> 
> I commited a fix.

@DanielJerolm, your commit fixes the problem. :+1:

Wishmaster-de commented:
I tried the build_doxygen_linux_x64*.sh scripts and i am getting an error concerning 'latex cannot be found'. I think there should be a dependency added to the 'doxygen-latex' package in [README](https://github.com/DanielJerolm/gpcc/blob/c2210aa538e2f84ce8f314621cc3a3bb010da9e8/README.md?plain=1#L48)
My doxygen version is 1.9.1 on a relative fresh Ubuntu Jammy installation.

DanielJerolm committed:
Get rid of doxygen-latex-dependency

There were two formula in the docoumentation that
seem to be rendered using latex. The formulas were of
no real value for the reader, so they were removed.

DanielJerolm commented:
There were two mathematical expressions in the documentation. It seems that doxygen renders them into bitmaps using latex. The formulas were removed, since they transported little information. Maybe this will remove the "latex cannot be found" issue.

DanielJerolm committed:
Refine wording

DanielJerolm committed:
Guess compiler and OS if not specified

DanielJerolm committed:
Test for unsupported OSAL config in productive env.

DanielJerolm committed:
Doc special test cases and related build options

DanielJerolm committed:
Move some docs into doc-folder

DanielJerolm committed:
Add unittest configuration for special non-tfc testcases

DanielJerolm committed:
Ensure guessing really takes placed when it is supposed to

DanielJerolm committed:
Update to googletest 1.16.0

DanielJerolm committed:
Update dox page "coding style"

DanielJerolm committed:
Wrap-up TODOs; Finalize

DanielJerolm commented:
The following will be done in subsequent, separate pull requests:

- Generation of doxygen scripts via CMake
- Adding a configuration using sanitizers
- Revision of compiler switches regarding warnings etc.

Wishmaster-de commented on scripts/execute_unittests_memcheck.sh:
```diff
@@ -0,0 +1,35 @@
+else
+
+  cd ../build_unittest/output
+  valgrind --tool=memcheck --leak-check=full --show-leak-kinds=definite,indirect,possible ./unittests $@
```
```suggestion
  valgrind --tool=memcheck --leak-check=full --show-leak-kinds=definite,indirect,possible ./unittests "$@"
```
Avoid word splitting in arguments.

Wishmaster-de commented on scripts/execute_unittests.sh:
```diff
@@ -0,0 +1,25 @@
+set -e
+
+cd ../build_unittest/output
+./unittests $@
```
```suggestion
./unittests "$@"
```
Avoid word splitting in arguments.

DanielJerolm committed:
Avoid word splitting in arguments

Co-authored-by: Wishmaster-de <74838617+Wishmaster-de@users.noreply.github.com>

DanielJerolm committed:
Avoid word splitting in arguments.

Co-authored-by: Wishmaster-de <74838617+Wishmaster-de@users.noreply.github.com>

DanielJerolm commented:
Interesting...
@Wishmaster-de: I accepted your suggestions via the github website. They turned into commits marked "Verified" but at the same time your review was dismissed.
